### PR TITLE
Add origin options to FormSpec labels

### LINKED ISSUE #7613: Formspec `label alignment
Would it be possible to add alignment to the formspec `label field so we can left|right|centre align text properly on ALL clients since gui_scale, fonts etc. differ greatly e.g.

### `label[<X>,<Y>;<label>;<alignment>]`
* The label formspec element displays the text set in `label`
  at the specified position.
* The text is displayed directly without automatic line breaking,
  so label should not be used for big text chunks.
* Aligntment can be nil for left aligned or set to "centre" to be
  centred or "right" to be right aligned on the X;Y position.

### PR DESCRIPTION
**Pull request by v-rob**
**Note**: This PR is closed due to bad implementation.  A new one will be worked on after #9755 is merged (or rejected for some reason).

This PR allows the `label` and `vertlabel` elements in FormSpecs to have origin set and custom newline spacing.  This closes #7613.  It adds three new styling options to `label` and `vertlabel`:

### From `lua_api.txt`:
* label, vertlabel
    * horigin - Sets horizontal origin in relation to label text. Possible values:
        * `left`: Left side of the bounding box
        * `center`: Center of the bounding box
        * `right`: Right side of the bounding box
        * `leftline` (vertlabel only): Left end of first column
        * `centerline` (vertlabel only): Center of first column
        * `rightline` (vertlabel only): Right end of first column
      Default `left` for label and `centerline` for vertlabel.
    * vorigin - Sets vertical origin in relation to label text. Possible values:
        * `top`: Top of the bounding box
        * `center`: Center of the bounding box
        * `bottom`: Bottom of the bounding box
        * `topline` (label only): Top of first line
        * `centerline` (label only): Center of first line
        * `bottomline` (label only): Bottom of first line
      Default `centerline` for label and `top` for vertlabel.
    * newline_spacing - number, sets space between lines/columns of text. If
      the number is prefixed with `*`, the spacing is set to a multiple of the
      label's font height or width, otherwise it's in coordinates. Default `0.5`.

In addition to these styling options, newlines have been added to `vertlabel` since it didn't previously support this.  Each newline moves the text to the right, unless, of course, it's negative.

A nice side effect of how it is implemented is that the `label` and `vertlabel` rects are as tall as the text height as opposed to a full coordinate as it is currently.  This also makes it so that it will work with really gigantic fonts as well as normal ones.

## Screenshot
A picture's worth a thousand words, so here you go:
image
The gray dots show where the label position is.

## To do

This PR is Ready to Review.

## How to test

<details><summary><b>It's <i>very</i> long</b></summary>

This is the above screenshot:
```lua
local function boxes(num_x, num_y)
	local boxes = ""
	for x = 0, num_x do
		for y = 0, num_y do
			if x % 2 == y % 2 then
				boxes = boxes .. "box["..x..","..y..";1,1;red]"
			end
		end
	end
	return boxes
end

local function point(x, y)
	return "box[".. x-0.05 ..",".. y-0.05 ..";0.1,0.1;gray]"
end

minetest.show_formspec("singleplayer", "formspec:test",
	"formspec_version[3]"..
	"size[20,12]"..
	boxes(19, 11)..
	---
	point(1, 1)..
	"style_type[label;horigin=left]".. -- Remove value for default label horigin: left
	"label[1,1;H-left]"..

	point(1, 2)..
	"style_type[label;horigin=center]"..
	"label[1,2;H-center]"..

	point(1, 3)..
	"style_type[label;horigin=right]"..
	"label[1,3;H-right]"..

	"box[2.95,0;0.1,12;gray]"..
	"box[0,6.95;3,0.1;gray]"..
	---
	point(1, 8)..
	"style_type[label;horigin=;newline_spacing=*1]"..
	"label[1,8;Newline 1 *1\nNewline 2 *1\nNewline 3 *1]"..
	
	point(1, 11)..
	"style_type[label;horigin=;newline_spacing=*-2]"..
	"label[1,11;Newline 1 *-2\nNewline 2 *-2\nNewline 3 *-2]"..
	---
	"style_type[label;newline_spacing=0.33]"..

	point(4, 1)..
	"style_type[label;vorigin=top]".. -- Remove value for default label vorigin: centerline
	"label[4,1;V-top\nV-top2\nV-top3]"..

	point(4, 3)..
	"style_type[label;vorigin=center]"..
	"label[4,3;V-center\nV-center2\nV-center3]"..

	point(4, 5)..
	"style_type[label;vorigin=bottom]"..
	"label[4,5;V-bottom\nV-bottom2\nV-bottom3]"..

	point(4, 6)..
	"style_type[label;vorigin=topline]"..
	"label[4,6;V-topline\nV-topline2\nV-topline3]"..

	point(4, 8)..
	"style_type[label;vorigin=centerline]"..
	"label[4,8;V-centerline\nV-centerline2\nV-centerline3]"..

	point(4, 10)..
	"style_type[label;vorigin=bottomline]"..
	"label[4,10;V-bottomline\nV-bottomline2\nV-bottomline3]"..

	"box[5.95,0;0.1,12;gray]"..
	---
	"style_type[label;newline_spacing=-0.33]"..

	point(7, 1)..
	"style_type[label;vorigin=top]"..
	"label[7,1;V-top\nV-top2\nV-top3]"..

	point(7, 3)..
	"style_type[label;vorigin=center]"..
	"label[7,3;V-center\nV-center2\nV-center3]"..

	point(7, 5)..
	"style_type[label;vorigin=bottom]"..
	"label[7,5;V-bottom\nV-bottom2\nV-bottom3]"..

	point(7, 6)..
	"style_type[label;vorigin=topline]"..
	"label[7,6;V-topline\nV-topline2\nV-topline3]"..

	point(7, 8)..
	"style_type[label;vorigin=centerline]"..
	"label[7,8;V-centerline\nV-centerline2\nV-centerline3]"..

	point(7, 10)..
	"style_type[label;vorigin=bottomline]"..
	"label[7,10;V-bottomline\nV-bottomline2\nV-bottomline3]"..

	"box[8.95,0;0.1,12;gray]"..
	---
	"style_type[label;horigin=]"..

	point(10, 2)..
	"style_type[label;vorigin=top]".. -- Remove value for default vertlabel vorigin: top
	"vertlabel[10,2;V|TOP]"..

	point(11, 2)..
	"style_type[label;vorigin=center]"..
	"vertlabel[11,2;V|CEN]"..

	point(12, 2)..
	"style_type[label;vorigin=bottom]"..
	"vertlabel[12,2;V|BTM]"..

	"box[9,3.95;11,0.1;gray]"..
	"box[15.95,0;0.1,4;gray]"..
	---
	point(17, 1)..
	"style_type[label;vorigin=;newline_spacing=*1]"..
	"vertlabel[17,1;\\\\N 1 *1\n\\\\N 2 *1\n\\\\N 3 *1]"..
	
	point(19, 1)..
	"style_type[label;vorigin=;newline_spacing=*-2]"..
	"vertlabel[19,1;\\\\N 1 *-2\n\\\\N 2 *-2\n\\\\N 3 *-2]"..
	---
	"style_type[label;newline_spacing=0.25]"..

	point(10, 5)..
	"style_type[label;horigin=left]".. -- Comment for default horigin: centerline
	"vertlabel[10,5;H|LFT\nH|LFT2\nH|LFT3]"..

	point(12, 5)..
	"style_type[label;horigin=center]"..
	"vertlabel[12,5;H|CEN\nH|CEN2\nH|CEN3]"..

	point(14, 5)..
	"style_type[label;horigin=right]"..
	"vertlabel[14,5;H|RGT\nH|RGT2\nH|RGT3]"..

	point(15, 5)..
	"style_type[label;horigin=leftline]"..
	"vertlabel[15,5;H|LFL\nH|LFL2\nH|LFL3]"..

	point(17, 5)..
	"style_type[label;horigin=centerline]"..
	"vertlabel[17,5;H|CTL\nH|CTL2\nH|CTL3]"..

	point(19, 5)..
	"style_type[label;horigin=rightline]"..
	"vertlabel[19,5;H|RTL\nH|RTL2\nH|RTL3]"..

	"box[9,7.95;11,0.1;gray]"..
	---
	"style_type[label;newline_spacing=-0.25]"..

	point(10, 9)..
	"style_type[label;horigin=left]"..
	"vertlabel[10,9;H|LFT\nH|LFT2\nH|LFT3]"..

	point(12, 9)..
	"style_type[label;horigin=center]"..
	"vertlabel[12,9;H|CEN\nH|CEN2\nH|CEN3]"..

	point(14, 9)..
	"style_type[label;horigin=right]"..
	"vertlabel[14,9;H|RGT\nH|RGT2\nH|RGT3]"..

	point(15, 9)..
	"style_type[label;horigin=leftline]"..
	"vertlabel[15,9;H|LFL\nH|LFL2\nH|LFL3]"..

	point(17, 9)..
	"style_type[label;horigin=centerline]"..
	"vertlabel[17,9;H|CTL\nH|CTL2\nH|CTL3]"..

	point(19, 9)..
	"style_type[label;horigin=rightline]"..
	"vertlabel[19,9;H|RTL\nH|RTL2\nH|RTL3]"
)
```
</details>

### CODE REVIEW

**Comment by v-rob** [comment]
Can this, along with #8665, #8652, and #8646 be added to the formspec merging list (https://github.com/minetest/minetest/projects/6)?

I know there are a lot of formspec PRs, but I think it's best for all formspec changes to be made in a single release since they require a client update for them to work properly.  Don't worry, I don't have *too* many left to do ;-)

**Comment by paramat** [comment]
Seems to be actively worked on so removing label.

**Comment by v-rob** [comment]
This PR is now Ready for Review

I've added support for `vertlabel`.  I've given them newline support as well just like `label`.  Note that this doesn't break anything: if you use a newline in a `vertlabel` before now, it will cut off the text in the rect, plus the fact that this all only works with real coordinates enabled.

File: doc/lua_api.txt
Comment (SmallJoker) [inline doc/lua_api.txt]:
```suggestion
* Newlines in a vertlabel add columns to the right of the starting text.
```

File: src/gui/guiFormSpecMenu.cpp
Comment (SmallJoker) [inline src/gui/guiFormSpecMenu.cpp]: Line needs breaking. Also superfluous parentheses and space in `((float) imgsize.Y)`.

**Review by SmallJoker — COMMENTED** [review]
Compared formspec screenshots before/after this PR - they look the same. The testing mod also displays the correct alignments.
Only a few stylistic change requests before the approval.

**Comment by ClobberXD** [comment]
There's a syntax error in the test code in the first post:

```lua
	"style_type[label;newline_spacing=0.3333;halign=center]"..
	"vertlabel[9,4;-VERTLABEL\n-WITH\n-0.3333\n-SPACING\n-COOL!]"..
	-- Note the trailing string concat (..) operator
)
```

**Comment by ClobberXD** [comment]
Tested using the test code in the OP, PR works as expected. :+1:

**Review by SmallJoker — DISMISSED** [review]

**Comment by v-rob** [comment]
I've rebased this PR as well as #9496, #8679, and #8676.  Blugh.

Both this PR and #8676 have one approval and have nothing left to do.

**Comment by kilbith** [comment]
Tested.

You have inverted `right/left` and `top/bottom` actually. For example `right aligned` should _extend_ to the left to keep the right side aligned. And so on...

Also this style only works with formspec version 3. Why does this need to be exclusive to it?

**Comment by rubenwardy** [comment]
The screenshot doesn't make sense to me - you don't give it a size, yet the position of the label text changes. A single-line size-less label should _never_ change position. When a size isn't specified, the label IGUIElement should be sized based on the countained text

**Comment by rubenwardy** [comment]
Oh I see - it's alignment in terms of the origin of the element, not the contents

**Comment by rubenwardy** [comment]
I suggest calling this "origin", as alignment in reference to text has another meaning

File: doc/lua_api.txt
Comment (rubenwardy) [inline doc/lua_api.txt:2649]: Please split this into align and origin. Align should set `setTextAlignment`. Origin should be used for the positioning logic

File: src/gui/guiFormSpecMenu.cpp
Comment (rubenwardy) [inline src/gui/guiFormSpecMenu.cpp:1735]: this should be `m_font->getDimension(wlabel_plain.c_str()).Height;` ?

Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:1735]: I seem to recall having trouble with this; I think it made the vertical rect too short for the text.  Then again, looking at the centered vertlabel in the screenshot, the rect appears a bit too tall.  I'll experiment around with it again.

I don't see how it breaks compat, though...

Comment (rubenwardy) [reply src/gui/guiFormSpecMenu.cpp:1735]: Oh sorry - what I mean is that it assumes there is only one line in the test

File: doc/lua_api.txt
Comment (v-rob) [reply doc/lua_api.txt:2649]: Renaming `align` to `origin` and then inverting would be fine with me.

I had no plans to make a separate `align` for multiple lines because I don't think it would be very useful (you don't know the width of the rect, so it makes it hard to know how to place it on the form), but I can add it if you want.

File: src/gui/guiFormSpecMenu.cpp
Comment (rubenwardy) [reply src/gui/guiFormSpecMenu.cpp:1735]: I suppose that it may not be worth doing anything too drastic here to support multiline labels well, as an area label/textblock is really needed for multiline labels

File: doc/lua_api.txt
Comment (rubenwardy) [reply doc/lua_api.txt:2649]: > you don't know the width of the rect, so it makes it hard to know how to place it on the form, and plus, it might confuse people

The code already supports alignment though? `e->setTextAlignment(halign, valign);`

Comment (rubenwardy) [reply doc/lua_api.txt:2649]: > then inverting would be fine with me

What you currently call align shouldn't be inverted - that's correct.

--------------------

Text alignment only really makes sense with multi-line labels, and multi-line labels are strange with the label[] element

**Comment by rubenwardy** [comment]
OK, a quick summary: currently, labels are positioned at the left middle of the first line. This PR changes is so that the positioning can be changed to not be the left-middle, it could be the center-bottom of the first line for example. The problems with this are:

1. `Align` is a name that is confusing, given both "Text Alignment" and the alignment in HUDs. Alignment in HUDs is relative to the entire element
2. Text alignment is automatically done as the mirror of `align`

**Comment by rubenwardy** [comment]
So, if we want to keep multi-line support well:

* Rename `alignment` to `origin`. Make the default `origin` be `left,baseline`.
* `horigin`: mirror horizontally. `|Left  `, ` Cen|ter `, `  right|`
* `vorigin`: mirror vertically
	* `top` - top of bounding box
	* `baseline` - baseline/middle of first line
	* `middle` - middle of bounding box
	* `bottom` - bottom of bounding box
* Text alignment can be set automatically as it is now

This makes it much easier to reuse when adding labels with explicit bounding boxes (ie: explicit sizes)

**Comment by v-rob** [comment]
> Also this style only works with formspec version 3. Why does this need to be exclusive to it?

Actually, it's just exclusive to `real_coordinates`, but I know that's a problem for you.  The reason why is that the way `real_coordinates` positions labels uses a different origin than the old system.  This new alignment is based on the `real_coordinates` version.

Making a non-real_coordinates version probably would work, but it would require separate defaults for each system, which is feasible.  I'll try it out after I fix everything else.

**Comment by rubenwardy** [comment]
> This makes it much easier to reuse when adding labels with explicit bounding boxes (ie: explicit sizes)

I'd like to add support for labels with explicit sizes in the future, which will make handling font stuff much easier. For example:

```
style_type[label;texthalign=center]
label[0,1;3,1;Hello\none\two]
```

**Comment by v-rob** [comment]
`baseline` vs `middle` is a good idea.  How come I didn't think of that?

I'm going to have `horigin` and `vorigin` like now instead of a combined `origin`.  Modders will probably want to just change one or the other in most cases, not both at the same time.  I don't know if you intended a combined one in the above post, but I want to be completely certain.

---

And I have to rebase yet again.  Ugh!

**Comment by rubenwardy** [comment]
> I'm going to have horigin and vorigin like now instead of a combined origin. Modders will probably want to just change one or the other in most cases, not both at the same time. I don't know if you intended a combined one in the above post, but I want to be completely certain.

I did mean horigin and vorigin. I was just being lazy!

**Comment by rubenwardy** [comment]
> baseline vs middle is a good idea. How come I didn't think of that?

Baseline isn't entirely correct, actually - baseline is far from the middle in most fonts. Idk of a better name

File: src/gui/guiFormSpecMenu.cpp
Comment (rubenwardy) [inline src/gui/guiFormSpecMenu.cpp:1732]: I'm getting compile errors here:

```
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp: In member function ‘void GUIFormSpecMenu::parseLabel(GUIFormSpecMenu::parserData*, const string&)’:
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:1732:48: error: cannot convert ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’} to ‘const wchar_t*’
 1732 |    first_char_base = m_font->getDimension(std::string(unescaped[0], 1)).Width / 2;
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~
      |                                                |
      |                                                std::string {aka std::__cxx11::basic_string<char>}
In file included from /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:38:
/usr/include/irrlicht/IGUIFont.h:58:61: note:   initializing argument 1 of ‘virtual irr::core::dimension2d<unsigned int> irr::gui::IGUIFont::getDimension(const wchar_t*) const’
   58 |  virtual core::dimension2d<u32> getDimension(const wchar_t* text) const = 0;
      |                                              ~~~~~~~~~~~~~~~^~~~
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp: In member function ‘void GUIFormSpecMenu::parseVertLabel(GUIFormSpecMenu::parserData*, const string&)’:
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:1866:33: error: invalid initialization of reference of type ‘const string&’ {aka ‘const std::__cxx11::basic_string<char>&’} from expression of type ‘std::__cxx11::basic_string<wchar_t>’
 1866 |     utf8_to_wide(unescape_string(lines[0]))));
      |                  ~~~~~~~~~~~~~~~^~~~~~~~~~
In file included from /home/ruben/dev/minetest/src/metadata.h:25,
                 from /home/ruben/dev/minetest/src/itemstackmetadata.h:22,
                 from /home/ruben/dev/minetest/src/inventory.h:24,
                 from /home/ruben/dev/minetest/src/inventorymanager.h:22,
                 from /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.h:27,
                 from /home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:26:
/home/ruben/dev/minetest/src/util/string.h:67:46: note: in passing argument 1 of ‘std::wstring utf8_to_wide(const string&)’
   67 | std::wstring utf8_to_wide(const std::string &input);
      |                           ~~~~~~~~~~~~~~~~~~~^~~~~
/home/ruben/dev/minetest/src/gui/guiFormSpecMenu.cpp:1874:51: error: no matching function for call to ‘max(float&, unsigned int)’
 1874 |       m_font->getDimension(lines[i].c_str()).Width);
```

Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:1732]: I'm aware of the compiler errors; due to the weird and unique way I compile Minetest, I have to push to GitHub before I can compile.  I'll notify you when I have it working.

File: src/gui/guiFormSpecMenu.cpp
Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:1724]: Actually, strangely enough, `stof` and `stoi` don't give errors; if the number is invalid, it returns `0`.  Try inputting an invalid number to anything that takes a number, e.g. any element position.  No element positioner does error checking: https://github.com/minetest/minetest/blob/master/src/gui/guiFormSpecMenu.cpp#L280-L299.  I don't know why it does this, but I'm kind of glad it does.

`stof("110%")` will return `110.0`; it searches up until the first invalid character.

Comment (SmallJoker) [reply src/gui/guiFormSpecMenu.cpp:1733]:
```C++
bool ok = sscanf(str_newline, "%f", &newline_spacing);
if (!ok)
	newline_spacing = 0;
newline_spacing += imgsize.Y;
```
This would be the sophisticated approach.

Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:1733]: I don't see what's the point of this when _every_ other formspec element uses `stoi` and `stof`, relying on the returning `0` behaviour.  If it gets changed, it should be changed for everything.

But that `sscanf` approach is pretty nice, all the same.

Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:1838]: I found the offenders.  It isn't using `std::stoi` or `std::stof` after all (I knew it seemed strange because these are C++11 which Minetest wasn't officially using until a little while ago, but tons of ancient code uses it) but `atoi` and `atof`:

https://github.com/minetest/minetest/blob/0e88176db8ce9f3fb631feb56bdd29d4ada054f5/src/util/string.h#L342-L391

**Comment by v-rob** [comment]
OK, so I've done some more work, and I've finally got everything down.  Look at the top for more details because I'm too lazy to write it all out here :)  It turned out to be a lot more than I expected.

---

As for old coordinate support:  The old labels and vertlabels just have too weird of rects for this to really be feasible.  For instance, this is `label`:

```C++
rect = core::rect<s32>(
	pos.X, pos.Y - m_btn_height,
	pos.X + m_font->getDimension(wlabel_plain.c_str()).Width,
	pos.Y + m_btn_height);

// snip

e->setTextAlignment(gui::EGUIA_UPPERLEFT, gui::EGUIA_CENTER);
```

First off, the Y position is offset (more than is even shown here), but the main problem is that vertical alignment is set to the center.  To emulate this behaviour while combining with the new alignment, a lot more compatibility code would be necessary, and I don't think it's justifiable for a now legacy system.  `vertlabel` would take a similar amount of code.

---

I'll take this time while I'm waiting for what I hope to be the last compile to explain some of my rationale:  Sized labels are a good idea, but this unsized way is much simpler for many things.  Alignment is good for `textarea` because these are better for displaying multiple lines of text.  Labels are better for simple descriptions.

If you look at sorcerykid's [polygraph](https://forum.minetest.net/viewtopic.php?t=20734), it has labels above the bars and dots and stuff.  This can't be perfectly centered in the old system, but it could be centered with either `align` or `origin`.  But `origin` is simpler: put the position right above the bar.  `align` has to be offset and have a size.  `align` might also be clipped for larger fonts or smaller screen sizes.  This is the real reason I made this how I did.

**Comment by v-rob** [comment]
Would it be good to add `checkbox` to this since it is similar?  I think `vorigin` would be a good idea, whereas `horigin` wouldn't be very useful.  It wouldn't require the `*line` options, only having a single line.  I can leave it for another PR as well, of course.

**Comment by SmallJoker** [comment]
@v-rob Support for `checkbox` should be done within its own PR if you'd like to add support for it. IMO it does not need aligning options due to its limited feature-set and relatively rare use.

> I suggest calling this "origin", as alignment in reference to text has another meaning

I believe `anchor` is also used in some places: top/center/bottom, left/center/right.

**Comment by v-rob** [comment]
I made one tiny change: `auto` newline spacing has been changed to `*number` as a multiple of font height or width.  Sorry for any inconvenience, but it's only a few lines change for more functionality.

File: doc/lua_api.txt
Comment (rubenwardy) [inline doc/lua_api.txt:2733]: I don't understand this comment?

File: doc/lua_api.txt
Comment (rubenwardy) [inline doc/lua_api.txt:2736]: I worry that the names of "xline" may be confusing, given it's more like the first character on the first line

**Comment by rubenwardy** [comment]
There's a bit of repetition here, not necessary a bad problem

It would be good to have the test formspec in `minimal`. The points are also clearer when red: 

image

File: doc/lua_api.txt
Comment (v-rob) [reply doc/lua_api.txt:2736]: I could change the `horigin` values to `*column` since the first character can be considered a column.  For `vertlabel`s, it actually is the center of the entire column, not just the first character.  Making different values for `label` and `vertlabel` would just make the docs too verbose, I think.

File: doc/lua_api.txt
Comment (v-rob) [reply doc/lua_api.txt:2733]: This was meant to help remove confusion by showing that `left` acts the same as `leftline` (same applies to `top` and `topline` for `vorigin`) when `newline_spacing` is positive, since the left side of the whole rect and the left side of the first line are exactly the same.  But it apparently adds more confusion; I can just remove it, unless you have a better way of putting it.

**Comment by sorcerykid** [comment]
Hmm, this seems way too complicated for label elements. Even being an avid developer with formspecs myself, I would honestly never use 90% of the options in this PR. As I recall, #7613 just asked for basic left, centere, and right alignment, I cannot imagine many use-cases where people require more sophisticated fine-control just to position labels above or beside a field. 

I get the impression the overall feature set of formspecs is very off-balance and lacks an agreed-upon roadmap. For example, I just saw a feature request for embedding any kind of content inside buttons. It's still not even possible to dynamically update formspecs in the client, yet apparently we'll be able to have buttons that can have entire mini-formspecs in them. That doesn't make any sense O_o

I seem to recall Rubenwardy had an issue opened awhile back for refactoring the entire formspec backend. I guess that's not happening now that the goal is just randomly adding cool and fun features.

**Comment by rubenwardy** [comment]
The extra features are required in order to retain compatibility and consistency. You could remove those and literally only have  left/center/right/legacy

There's still no consensus on a refactor, I received no response to my suggestion

**Comment by Df458** [comment]
>  It's still not even possible to dynamically update formspecs in the client, yet apparently we'll be able to have buttons that can have entire mini-formspecs in them. That doesn't make any sense O_o

Sure it does. It's an easy feature that's impossible for mods to properly hack in with the current API, and the necessary backend work lays the API groundwork for a larger and more important feature--allowing tabs to have 'pages' of content instead of forcing a re-send every tab change.

Dynamically updating formspecs is super important, but it requires significant design and implementation effort to get in and is more in the 'concept phase' than anything else. Expecting people to be working on something that isn't actually hashed out fully is a great way to make bad API decisions--something that Minetest suffers greatly from as-is.

**Comment by v-rob** [comment]
I personally consider the `*line` options more useful; that's what the original PR did.  I realize that some people might consider the non-`line` options useful though.  The `line` for the first char isn't really that useful, but wasn't hard to implement.  It can be used to line up a `label` and a `vertlabel` at the same position so that they have the same first character.  I can still remove that if wanted.

**Comment by v-rob** [comment]
I think I should probably clarify my rationale for making the `line` options:  Labels are primarily for putting text next to, above, or somewhere in relation to an element in order to give it more meaning.  They aren't intended as large blocks of text; that's `textarea`'s job (and see https://github.com/minetest/minetest/pull/9755, which is a much smaller PR besides), but as small descriptors.  Now, let's look at the `real_coordinates` screenshot:



See the label with newlines over the buttons?  Try aligning that without the `line` options.  And what if each label has to be above the button with `topline`?  I suppose that separate labels can be used, but that takes more work, and it's impossible to combine when making the labels have a `newline_spacing` as a multiple of font size.

I realize that non-`line` options are useful for making a pseudo-textarea without the scrolling, word wrap, or highlighting, which can be used for, e.g., an overlay over a button, but that doesn't discount the usefulness of the `line`s.

I can remove the `line` options if others are insistent, but I find them useful, and they add barely any more code.  The non-`line` options use more code, actually.

The character alignment variation of the `line` options isn't really useful -- that's something I think I will remove.  I only added that because I could, and that is a bad reason.

**Comment by v-rob** [comment]
I was mulling over this PR a bit yesterday, and I think I've revised my position on whether or not this should be merged.  It seems pretty certain that sized labels with `halign` and `valign` will be made at some point, and I suppose it would be redundant to have two similar things, even though they have subtle differences.  It would probably end up as confusing for modders and not worth the extra code in the engine.

As a result, I think I will close this PR and create one for aligned sized labels after the very simple #9755 is merged to minimize conflicts.  It would probably have `newline_spacing` as well, but it would not include `vertlabel`s whatsoever since that can be emulated with `label` with a newline after every character and a `newline_spacing` of `*1`.  I wouldn't include `*line` options because they wouldn't work in that case well anyway.

I won't act without feedback, though, so I would appreciate any anyone could give me.

**Comment by sorcerykid** [comment]
Thanks for being receptive to my feedback v-rob. And sorry if I kind of lashed out with so many criticisms before. I value the work you do, and this PR (at least in its earlier state) had the exact feature-set that I'd dreamed about for years. So it's not for lack of appreciation of the concept. It's moreso that I just feel it's important that formspecs have at least remain balanced as possible when it comes to capabilities vs. complexity. I tend to feel that a generic API that solves at least the most common use-cases, while still accommodating edge-cases via documented workarounds, is usually the most elegant solution. Hope that makes sense :)

**Pull request closed without merging**
